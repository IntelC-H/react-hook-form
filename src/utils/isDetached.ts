import { Ref } from '../types';

export default function isDetached(element: Ref): boolean {
  // falsy elements are detached (probably a null parent)
  if (!element) return true;
  // Non HTML Elements should never be considered detached
  if (!(element instanceof HTMLElement)) return false;
  // If we can find our way up to a document node, we assume we are still attached!
  if (element.nodeType === Node.DOCUMENT_NODE) return false;
  // Recursive call
  return isDetached(element.parentNode);
}
