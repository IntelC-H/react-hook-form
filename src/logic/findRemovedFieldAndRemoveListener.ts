import removeAllEventListeners from './removeAllEventListeners';
import isRadioInput from '../utils/isRadioInput';
import isCheckBoxInput from '../utils/isCheckBoxInput';
import isDetached from '../utils/isDetached';
import isArray from '../utils/isArray';
import { Field, FieldRefs, FieldValues } from '../types';

export default function findRemovedFieldAndRemoveListener<
  FormValues extends FieldValues
>(
  fields: FieldRefs<FormValues>,
  handleChange: ({ type, target }: Event) => Promise<void | boolean>,
  field: Field,
  forceDelete?: boolean,
): void {
  if (!field) {
    return;
  }

  const {
    ref,
    ref: { name, type },
    mutationWatcher,
  } = field;

  if (!type) {
    delete fields[name];
    return;
  }

  const fieldValue = fields[name];

  if ((isRadioInput(ref) || isCheckBoxInput(ref)) && fieldValue) {
    const { options } = fieldValue;

    if (isArray(options) && options.length) {
      options.forEach(({ ref, mutationWatcher }, index): void => {
        if ((ref && isDetached(ref)) || forceDelete) {
          removeAllEventListeners(ref, handleChange);

          if (mutationWatcher) {
            mutationWatcher.disconnect();
          }

          options.splice(index, 1);
        }
      });

      if (options && !options.length) {
        delete fields[name];
      }
    } else {
      delete fields[name];
    }
  } else if (isDetached(ref) || forceDelete) {
    removeAllEventListeners(ref, handleChange);

    if (mutationWatcher) {
      mutationWatcher.disconnect();
    }
    delete fields[name];
  }
}
