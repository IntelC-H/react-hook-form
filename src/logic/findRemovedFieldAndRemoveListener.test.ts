import findRemovedFieldAndRemoveListener from './findRemovedFieldAndRemoveListener';
import isDetached from '../utils/isDetached';

jest.mock('./removeAllEventListeners');
jest.mock('../utils/isDetached');

describe('findMissDomAndClean', () => {
  beforeEach(() => {
    jest.resetAllMocks();

    (isDetached as any).mockImplementation(() => {
      return true;
    });
  });

  it('should return default fields value if nothing matches', () => {
    document.body.contains = jest.fn(() => true);
    const fields = {
      test: 'test',
    };
    expect(
      findRemovedFieldAndRemoveListener(
        fields as any,
        () => {},
        {
          ref: { name: 'bill', type: 'radio' },
        },
      ),
    ).toEqual(undefined);
  });

  it('should remove options completely if option found and no option left', () => {
    document.body.contains = jest.fn(() => false);

    const ref = document.createElement('input');
    ref.setAttribute('name', 'test');
    ref.setAttribute('type', 'radio');

    const disconnect = jest.fn();
    const fields = {
      test: {
        name: 'test',
        ref: {},
        options: [
          {
            ref: 'test',
            mutationWatcher: {
              disconnect,
            },
          },
        ],
      },
    };
    findRemovedFieldAndRemoveListener(fields, () => {}, {
      ref: { name: 'test', type: 'radio' },
      options: [
        {
          mutationWatcher: {
            disconnect,
          },
          ref,
        },
      ],
    });

    expect(fields).toEqual({});
  });

  it('should remove none radio field when found', () => {
    const ref = document.createElement('input');
    ref.setAttribute('name', 'test');
    ref.setAttribute('type', 'radio');
    document.body.contains = jest.fn(() => false);
    const disconnect = jest.fn();
    const fields = {
      test: {
        name: 'test',
        ref: {},
        mutationWatcher: {
          disconnect,
        },
      },
      test1: {
        name: 'test',
        ref: {},
      },
    };

    findRemovedFieldAndRemoveListener(fields, () => {}, {
      ref,
      mutationWatcher: {
        disconnect,
      },
    });

    expect(fields).toMatchSnapshot();
  });

  it('should return undefined when empty ref', () => {
    const fields = {
      test: 'test',
    };
    expect(
      findRemovedFieldAndRemoveListener(
        fields as any,
        () => {},
        {} as any,
      ),
    ).toEqual(undefined);
  });

  it('should work for radio type input', () => {
    const ref = document.createElement('input');
    ref.setAttribute('name', 'test');
    ref.setAttribute('type', 'radio');
    document.body.contains = jest.fn(() => false);
    const disconnect = jest.fn();
    const fields = {
      test: {
        name: 'test',
        ref: {},
        mutationWatcher: {
          disconnect,
        },
      },
      test1: {
        name: 'test',
        ref: {
          type: 'radio',
        },
      },
    };

    expect(
      findRemovedFieldAndRemoveListener(fields, () => {}, {
        ref: { name: 'test', type: 'radio' },
        options: [{ ref }],
        mutationWatcher: {
          disconnect,
        },
      }),
    ).toMatchSnapshot();
  });

  it('should not remove event listener when type is not Element', () => {
    (isDetached as any).mockImplementation(() => {
      return false;
    });
    document.body.contains = jest.fn(() => false);

    const disconnect = jest.fn();
    const fields = {
      test: {
        name: 'test',
        type: 'radio',
        ref: {},
        options: [
          {
            ref: 'test',
            mutationWatcher: {
              disconnect,
            },
          },
        ],
      },
    };

    findRemovedFieldAndRemoveListener(fields, () => {}, {
      ref: { name: 'test', type: 'text' },
      options: [
        {
          mutationWatcher: {
            disconnect,
          },
          ref: {},
        },
      ],
    });

    expect(fields).toMatchSnapshot();

    expect(
      findRemovedFieldAndRemoveListener(fields, () => {}, {
        ref: { name: 'test', type: 'text' },
      }),
    ).toMatchSnapshot();
  });

  it('should return undefined when field is not found', () => {
    expect(
      findRemovedFieldAndRemoveListener({}, undefined, undefined as any, false),
    ).toBeUndefined();
  });

  it('should remove options when force delete is set to true', () => {
    (isDetached as any).mockImplementation(() => {
      return false;
    });

    document.body.contains = jest.fn(() => false);

    const ref = document.createElement('input');
    ref.setAttribute('name', 'test');
    ref.setAttribute('type', 'radio');

    const disconnect = jest.fn();
    const fields = {
      test: {
        name: 'test',
        ref: {},
        options: [],
      },
    };
    findRemovedFieldAndRemoveListener(
      fields,
      () => {},
      {
        ref: { name: 'test', type: 'radio' },
        options: [
          {
            mutationWatcher: {
              disconnect,
            },
            ref,
          },
        ],
      },
      true,
    );

    expect(fields).toEqual({});
  });
});
