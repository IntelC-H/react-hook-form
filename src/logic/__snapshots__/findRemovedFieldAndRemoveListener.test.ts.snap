// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`findMissDomAndClean should remove none radio field when found 1`] = `
Object {
  "test1": Object {
    "name": "test",
    "ref": Object {},
  },
}
`;

exports[`findMissDomAndClean should remove options completely if option found and no option left 1`] = `Object {}`;

exports[`findMissDomAndClean should work for radio type input 1`] = `
Object {
  "test1": Object {
    "name": "test",
    "ref": Object {
      "type": "radio",
    },
  },
}
`;
